src/Tolerance/Operation/Runner/RetryPromiseOperationRunner.php

```diff
@@ -105,6 +105,11 @@ private function onTerminate(PromiseOperation $operation, ThrowableCatcherVoter
                     return $fulfilled ? new FulfilledPromise($value) : new RejectedPromise($value);
                 }
 
+                // If it is a php http Promise, use php http's objects.
+                if ($promise instanceof \Http\Promise\Promise) {
+                    return $fulfilled ? new \Http\Client\Promise\HttpFulfilledPromise($value) : new \Http\Client\Promise\HttpRejectedPromise($value);
```

Comment:
I think there is some confusion here: `$promise` here is the same as you call the on terminate callback on. Why? Since `RetryPromiseOperationRunner::run` returns a [promise](https://github.com/Tolerance/Tolerance/pull/91/files#diff-9a5487d0f4ea328353fbbf8c40828a7dR40) which receives a promise in it's callback, `$value` itself in this case will be a promise, likely unresolved.

Why do you need to specifically set Fulfilled/Rejected promises here? It's useful when you already know the resolution which actually you don't: `$promise` is fulfilled, but there is no guarantee that `$value` is too, so you will have to call `->wait()` here as @joelwurtz pointed, but in that case you will loose real async.

But I would advise reviewing how promises work, seems to be a bit complicated to me. There are a few examples in PHP-HTTP documentation, Guzzle is a bit more sophisticated, but basically the same, you might want to check it's documentation out too.